### DISCUSSION TITLE
Is already Groonga 2038 year problem ready on 32bit arch??

### DISCUSSION THREAD

**Discussion by kenhys** [groonga/groonga#1698] [Q&A] [answered]
# Question

Is already Groonga 2038 year problem ready on 32bit arch?
I want to know whether  already created database is accessible after upgrading to 64bit time_t aware (rebuilt) groonga or not.

* grn_time_t is directly use time_t so if 32bit time_t environment, it may be problematic.
https://github.com/groonga/groonga/blob/main/include/groonga/portability.h#L210

* last_modified field (object header) is fixed to uint32_t, it is not problematic until 2106.
https://github.com/groonga/groonga/blob/main/lib/db.c#L2970
 
* Time column is int64_t, so not affected.
https://github.com/groonga/groonga/blob/main/lib/db.c#L15602

I assume that:
* even though after upgraded to rebuilt groonga on 64bit aware time_t transition environment, no need to re-create existing database for a while.
* after problematic 2038 year timestamp, internal timestamp also not affected.

Does it correct?

# Background

In Debian project, 64bit time_t transition is ongoing.

64-bit time_t transition in progress
https://lists.debian.org/debian-devel-announce/2024/02/msg00000.html

**Comment by kou** [comment]
>   * grn_time_t is directly use time_t so if 32bit time_t environment, it may be problematic.
>     https://github.com/groonga/groonga/blob/main/include/groonga/portability.h#L210

Debian will make `time_t` a 64bit value on 32bit environment, right?
If so, this is not a problem. We can treat `time_t` value as a 64bit value.
 
>   * last_modified field (object header) is fixed to uint32_t, it is problematic.
>     https://github.com/groonga/groonga/blob/main/lib/db.c#L2970

Ah, we need to change it:

```diff
diff --git a/lib/grn_io.h b/lib/grn_io.h
index 0acfa4a40..8c03911fa 100644
--- a/lib/grn_io.h
+++ b/lib/grn_io.h
@@ -89,7 +89,7 @@ struct _grn_io_header {
   uint32_t lock;
   uint64_t curr_size;
   uint32_t segment_tail;
-  uint32_t last_modified;
+  uint64_t last_modified;
 };
 
 struct _grn_io {
```

But it makes `struct _grn_io_header` 68 bytes... It's larger than https://github.com/groonga/groonga/blob/051e4361340ce6f4d2fecf26f250fa6df930be11/lib/io.c#L77 ...

It breaks DB compatibility. We need to think about this carefully...

>   * Time column is int64_t, so not affected.
>     https://github.com/groonga/groonga/blob/main/lib/db.c#L15602

But its unit is microsecond. Is it OK? 
 
>   * even though after upgraded to rebuilt groonga on 64bit aware time_t transition environment, no need to re-create existing database for a while.

I think so.

>   * after problematic 2038 year timestamp, internal timestamp also not affected.

I think so.

**Reply by kou** [reply, marked as answer]
> > * last_modified field (object header) is fixed to uint32_t, it is problematic.
> >   https://github.com/groonga/groonga/blob/main/lib/db.c#L2970
> 
> Ah, we need to change it:
> 
> ```diff
> diff --git a/lib/grn_io.h b/lib/grn_io.h
> index 0acfa4a40..8c03911fa 100644
> --- a/lib/grn_io.h
> +++ b/lib/grn_io.h
> @@ -89,7 +89,7 @@ struct _grn_io_header {
>    uint32_t lock;
>    uint64_t curr_size;
>    uint32_t segment_tail;
> -  uint32_t last_modified;
> +  uint64_t last_modified;
>  };
>  
>  struct _grn_io {
> ```
> 
> But it makes `struct _grn_io_header` 68 bytes... It's larger than
> 
> https://github.com/groonga/groonga/blob/051e4361340ce6f4d2fecf26f250fa6df930be11/lib/io.c#L77
> ...
> 
> It breaks DB compatibility. We need to think about this carefully...

It's not correct. It's `unit32_t`. So we can keep using it after 2038.

**Comment by kenhys** [comment]
FYI:

If you keep 32bit time_t environment, and continue to use it without fix after problematic 2038 timestamp, groonga will crash when updating database.

This issue will be solved by compiling groonga with abi=+time64 feature on 32bit time_t environment explicitly. (at least, it will be applied for such platforms. e.g armel, armhf and so on)

In Debian Project, above issue will be fixed in 64bit time_t transition process.
